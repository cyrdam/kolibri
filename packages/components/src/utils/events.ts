// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
