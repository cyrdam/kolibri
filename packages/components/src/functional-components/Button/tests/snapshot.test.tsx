import { h } from '@stencil/core';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import KolButtonFc from '../Button';

describe('KolButtonFc', () => {
	it('should render correctly', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolButtonFc label="Test Button" />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('KOL-BUTTON-WC');
		expect(page.root?.getAttribute('_label')).toContain('Test Button');
	});

	it('should render with custom class', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolButtonFc label="Test Button" class="custom-class" />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.className).toContain('custom-class');
	});

	it('should handle click event', async () => {
		const onClick = jest.fn();
		const page = await renderFunctionalComponentToSpecPage(() => <KolButtonFc label="Test Button" onClick={onClick} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.getAttribute('_on')).toBeDefined();
	});

	it('should render with icons', async () => {
		const icons = { left: 'icon-left', right: 'icon-right' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolButtonFc label="Test Button" icons={icons} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?._icons).toEqual(icons);
	});

	it('should hide label when hideLabel is true', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolButtonFc label="Test Button" hideLabel />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.getAttribute('_hideLabel')).toBeDefined();
	});
});
