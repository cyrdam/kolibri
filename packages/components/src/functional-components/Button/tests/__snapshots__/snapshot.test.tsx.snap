// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolButtonFc should handle click event 1`] = `
<kol-button-wc
  _label="Test Button"
  tabindex="0"
/>
`;

exports[`KolButtonFc should hide label when hideLabel is true 1`] = `
<kol-button-wc
  _hidelabel=""
  _label="Test Button"
  tabindex="0"
/>
`;

exports[`KolButtonFc should render correctly 1`] = `
<kol-button-wc
  _label="Test Button"
  tabindex="0"
/>
`;

exports[`KolButtonFc should render with custom class 1`] = `
<kol-button-wc
  _label="Test Button"
  class="custom-class"
  tabindex="0"
/>
`;

exports[`KolButtonFc should render with icons 1`] = `
<kol-button-wc
  _label="Test Button"
  tabindex="0"
/>
`;
